### PR TITLE
logReader: manually commit offset when using kafka logConsumer

### PR DESCRIPTION
**Pull request by welansari**
Kafka log consumer currently auto-commits messages after they are consumed. This is an issue as it means we loose messages after restart or crash even if their processing didn't finish.

Like we do with other logConsumer implementations, we now manually store offsets after the end of each batch. The auto-commit mechanism commits offsets stored locally after each 5 seconds.

Contrary to the BackbeatConsumer, there is no risk of offsets being committed in disorder as only one batch is processed at a time, and a partition is only assigned to a single instance of the QueuePopulator.

Issue: BB-698

### CODE REVIEW

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 1]
```diff
@@ -20,17 +20,19 @@ class LogConsumer {
...
         this._consumerGroupId = consumerGroupId;
+        this._topicPartition = null;
         this._log = logger;
-        this._offsets = {};
     }
 
     /**
      * Get partition offsets
-     * @returns {string} stored partition offsets
+     * Offsets are stored in kafka are not managed
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]: typo?
```suggestion
     * Offsets are stored in kafka and not managed
```

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 2]
```diff
@@ -42,13 +44,17 @@ class LogConsumer {
     setup(done) {
         // partition offsets will be managed by kafka
         const consumerParams = {
-            'enable.auto.offset.store': true,
+            // Manually manage storing offsets to ensure they are only stored
+            // after the batch processing is fully completed.
+            'enable.auto.offset.store': false,
+            // Default auto-commit interval is 5 seconds
+            'enable.auto.commit': true,
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:61]: with `auto-commit`, we will still keep a window (5 seconds) of uncertainty:
- is it the same with backbeatconsumer? I don't see the flag there...
- in backbeatConsume, I think we have code to ensure we "flush" the commited offset when needed : should we do the same here?

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 3]
```diff
@@ -42,13 +44,17 @@ class LogConsumer {
     setup(done) {
         // partition offsets will be managed by kafka
         const consumerParams = {
-            'enable.auto.offset.store': true,
+            // Manually manage storing offsets to ensure they are only stored
+            // after the batch processing is fully completed.
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:58]: bonus issue: this will raise the issues we had (and fixed) in BackbeatConsumer around handling of rebalance, i.e. the need to delay the rebalance until the batch has been completed and offsets commited accordingly, and the need to detect/handle "slow tasks" which could block all further processing... It may be done in another PR (or maybe we could reuse BackbeatConsumer for KafkaLogConsumer?), but we probably can't go to prod without that kind of things...

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 4]
```diff
@@ -155,16 +180,39 @@ class LogConsumer {
...
-                return next(null, { log: this._listRecordStream, tailable: false });
+        ], err =>{
+            if (err) {
+                return cb(err);
             }
-        ], (err, res) => cb(err, res[3]));
+            // ending and returning the stream
+            this._listRecordStream.end();
+            return cb(null, { log: this._listRecordStream, tailable: false });
+        });
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]: nit: i think the original flow was better, as it managed the callback (`cb`) in a single place, with no added condition (`if`) to handle error... or was there an issue? (we can also do `res?.[3]` if res could be null)

File: lib/queuePopulator/LogReader.js [hunk 5]
```diff
@@ -712,25 +712,50 @@ class LogReader {
...
-                this._metricsHandler.logReadOffset(this.getMetricLabels(), this.logOffset);
-                this._metricsHandler.logSize(this.getMetricLabels(), logSize);
+                this._updateMetrics(logRes.info);
             }
 
             return this._writeLogOffset(logger, done);
         }
+
+        // Handle offset managed externally (e.g., Kafka)
+        if (!this.isOffsetManaged() && this.logConsumer.storeOffsets) {
```
Comment (francoisferrand) [inline lib/queuePopulator/LogReader.js]: When `!this.isOffsetManaged()`, then `storeOffsets` should always be defined, right?
If this is the case, then this option is redundant (an assertion in constructor is enough)

This would allow a much nicer flow:
```
    if (!this.isOffsetManaged()) {
        this.logConsumer.storeOffsets();
    } else if (this._shouldUpdateLogOffset(logRes, nextLogOffset)) {
        ...
    }
```

Comment (welansari) [reply lib/queuePopulator/LogReader.js]: > When !this.isOffsetManaged(), then storeOffsets should always be defined, right?

Currently yes, the only logReader for which we don't manage offsets is the Kafka logReader.

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js (hunk 2 above)
Comment (welansari) [reply lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:61]: > is it the same with backbeatconsumer? I don't see the flag there...

Yes, we have the same in `BackbeatConsumer`, the option is enabled by default i just decided to included it here for clarity.

> in backbeatConsume, I think we have code to ensure we "flush" the commited offset when needed : should we do the same here?

Yes, i should handle rebalance to avoid replay cases

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 6]
```diff
@@ -136,8 +219,20 @@ class LogConsumer {
...
+                topicPartition.set(`${message.topic}-${message.partition}`, {
+                    topic: message.topic,
+                    partition: message.partition,
+                    offset: message.offset + 1, // next offset to commit
+                });
+                // writing consumed messages to the stream
+                this._listRecordStream.write(message);
+            });
+            // format offsets for commit
+            this._topicPartition = Array.from(topicPartition.values());
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:273]: Should we not `merge` this map with the existing `this._topicPartition` ?
e.g. in `_consumeKafkaMessages` is called multiple times before `storeOffsets()` is called

(maybe this does not happen, but it is not clear from the prototype of LogConsumer that it MUST work this way...)

In that case, may be better to keep a map in LogConsumer, and build the array only as a temporary varialbe in storeOffsets

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 7]
```diff
@@ -42,70 +46,141 @@ class LogConsumer {
...
+     * @param {kafka.KafkaError} err Rebalance event
+     * @param {TopicPartition[]} assignment List of (un)assigned partitions
+     * @returns {void}
      */
-    _storeCurrentOffsets(cb) {
-        this._consumer.committed(5000, (err, topicPartitions) => {
-            if (err) {
-                this._log.error('Error while getting offsets', {
-                    method: 'LogConsumer._storeCurrentOffsets',
+    _onRebalance(err, assignment) {
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:119]: should this update metrics: `KafkaBacklogMetrics.onRebalance(this._topic, this._groupId, status)` ?
(or is that metrics specific to BackbeatConsumer)

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 8]
```diff
@@ -42,70 +46,141 @@ class LogConsumer {
...
+                        method: 'LogConsumer._onRebalance',
+                        topic: this._topic,
+                        consumerGroupId: this._consumerGroupId,
+                    });
+                    clearInterval(rebalanceTimeoutInterval);
+                    this._consumer.disconnect();
+                    return;
+                }
+                // wait for consumed messages to finish being processed
+                if (this._topicPartition && this._topicPartition.length > 0) {
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]: instead of polling, can't we trigger this callback when the topic gets empty?
similar to BackbeatConsumer, we could create a `drain` callback which is called when `_topicPartition` is reset ; and set it when needed in rebalance.

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 9]
```diff
@@ -153,18 +248,80 @@ class LogConsumer {
...
+                        return null;
+                    }
+                }
+                return next();
+            },
+            next => {
+                if (this._consumer?.isConnected()) {
+                    this._consumer.disconnect();
+                    this._consumer.once('disconnected', () => next());
+                    return null;
```
Comment (francoisferrand) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]: should we not do it the other way, to avoid race conditions?
(node is thread safe, so probably fine as is... but could change with futurechanges...)
```suggestion                  
                    this._consumer.once('disconnected', () => next());
                    this._consumer.disconnect();
                    return null;
```

(unless the semantics is that setting `once('disconnected')` callback will immediately invoke the callback if the consumer is alreadyd disconnected)

File: lib/queuePopulator/LogReader.js [hunk 10]
```diff
@@ -111,10 +111,18 @@ class LogReader {
      * @return {undefined}
      */
     close(cb) {
-        async.each(
-            Object.values(this._producers),
-            (producer, done) => producer.close(done),
-            cb);
+        async.parallel([
```
Comment (francoisferrand) [inline lib/queuePopulator/LogReader.js]: should we not first (in series!) stop the consumer -waiting for all 'current' messages to have been consumer- then stop the producers?

File: tests/unit/lib/queuePopulator/LogReader.spec.js [hunk 11]
```diff
@@ -599,30 +599,34 @@ describe('LogReader', () => {
...
                 shouldSave: false,
             },
             {
                 desc: 'should not save the offset if the offset did not change',
                 managed: true,
+                hasStoreOffsetsFn: false,
                 currentOffset: 0,
                 nextLogOffset: 0,
                 shouldSave: false,
             },
```
Comment (francoisferrand) [inline tests/unit/lib/queuePopulator/LogReader.spec.js:622]: should we really remove that test:
- that would mean we changed the behavior, which seems weird?
- the other tests seem to have similar conditions ("offsets not managed"), so this not a corner case we dropped
- if we remove it from here, we should probably create a new one (or multiple, if offset are/aren't managed)

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js (hunk 6 above)
Comment (welansari) [reply lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:273]: The queuePopulator doesn't support parallel processing of batches, only a single batch of messages is consumed, processed and then committed at a time. I don't see a need to support that currently.

I enforced this by not allowing consumption of new messages when a batch is in progress.

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js (hunk 7 above)
Comment (welansari) [reply lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:119]: No the metric is not specific to BackbeatConsumer, i added it

File: tests/unit/lib/queuePopulator/LogReader.spec.js (hunk 11 above)
Comment (welansari) [reply tests/unit/lib/queuePopulator/LogReader.spec.js:622]: The behaviour did change for the non managed logConsumer (Kafka) and it's not weird, the whole point of this PR is to now handle how/when the offsets get committed. Before, we used to do nothing and the offsets would be committed as soon as messages were consumed, now we do it manually after each batch of messages finished processing.

**Review by francoisferrand — APPROVED** [review]

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 12]
```diff
@@ -42,70 +49,168 @@ class LogConsumer {
...
+            if (err.code === kafka.CODES.ERRORS.ERR__NO_OFFSET) {
+                return undefined;
             }
-            return cb();
-        }, 2000);
+            this._log.error('Error committing offsets to kafka', {
+                method: 'LogConsumer._onOffsetCommit',
+                errorCode: err,
+                topicPartitions,
+                groupId: this._groupId
```
Comment (ghost) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]: I'm not sure this is defined in this class?

```suggestion
                groupId: this._consumerGroupId
```

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 13]
```diff
@@ -42,70 +49,168 @@ class LogConsumer {
...
+            return undefined;
+        }
+        this._pendingCommit = false;
+        if (!this._hasUnprocessedMessages()) {
+            this.emit('drained');
+        }
+        this._log.debug('Offsets committed', {
+            method: 'LogConsumer._onOffsetCommit',
+            topicPartitions,
+            groupId: this._groupId
```
Comment (ghost) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]:
```suggestion
            groupId: this._consumerGroupId
```

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 14]
```diff
@@ -42,70 +49,168 @@ class LogConsumer {
...
+                method: 'LogConsumer._onOffsetCommit',
+                errorCode: err,
+                topicPartitions,
+                groupId: this._groupId
+            });
+            return undefined;
+        }
+        this._pendingCommit = false;
+        if (!this._hasUnprocessedMessages()) {
+            this.emit('drained');
```
Comment (ghost) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]:
```suggestion
            this.emit(constants.unassignStatus.DRAINED);
```

(same at other places)

File: lib/queuePopulator/KafkaLogConsumer/LogConsumer.js [hunk 15]
```diff
@@ -153,18 +281,84 @@ class LogConsumer {
...
+            method: 'LogConsumer.storeOffsets',
+            consumerGroupId: this._consumerGroupId,
+            topicPartition: this._topicPartition,
+        });
+        this._pendingCommit = true;
+        this._topicPartition = null;
+    }
+
+    _hasUnprocessedMessages() {
+        return this._topicPartition && this._topicPartition.length > 0;
```
Comment (ghost) [inline lib/queuePopulator/KafkaLogConsumer/LogConsumer.js]:
```suggestion
        return this._topicPartition?.length > 0;
```

File: lib/queuePopulator/LogReader.js [hunk 16]
```diff
@@ -777,6 +807,13 @@ class LogReader {
         return this._batchTimedOut;
     }
 
+    isLogConsumerReady() {
+        if (this.logConsumer?.isReady) {
+            return this.logConsumer.isReady();
+        }
+        return true;
```
Comment (ghost) [inline lib/queuePopulator/LogReader.js]:
```suggestion
        return this.logConsumer?.isReady?.() ?? true;
```

(just a personal preference, not blocking)

**Review by ghost — APPROVED** [review]
Nothing blocking to me, the first comment should probably be addressed though, as I think it will always log undefined otherwise
